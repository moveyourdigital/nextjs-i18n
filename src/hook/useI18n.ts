import { useRouter } from "next/router"
import { ReactNode, ReactText, isValidElement, cloneElement } from "react"

export type I18nLocale = Exclude<string, "count" | "params">
export type I18nParam = ReactText | ReactNode
export type I18nSingleValue = ReactText
export type I18nPluralValue = { [n: number]: ReactText; n?: ReactText }
export type I18nParams = Record<string, I18nParam> & { count?: number }

export type TranslateFn = (
  locales: Record<I18nLocale, I18nSingleValue | I18nPluralValue>,
  args: {
    params?: I18nParams
    count?: number
  },
) => ReactText | ReactNode

const TAG_REGEXP = /<([a-z0-9_-]+)\b[^>]*>(.*?)<\/\1>/gi
const STR_REGEXP = /{{([a-z0-9_-]+)\s*(.*?)}}/gi

function useI18n() {
  const { locale, locales, defaultLocale } = useRouter()

  const getBest = (
    locales: Record<I18nLocale, I18nSingleValue | I18nPluralValue>,
  ) => {
    if (locale && locale in locales) {
      return locales[locale]
    }

    if (defaultLocale && defaultLocale in locales) {
      return locales[defaultLocale]
    }

    throw new Error(`No string defined for ${defaultLocale}`)
  }

  const getValue = (
    locales: Record<I18nLocale, I18nSingleValue | I18nPluralValue>,
    count?: number,
  ) => {
    const value = getBest(locales)
    if (typeof value === "object") {
      if (count in value) {
        return value[count]
      } else {
        return value["n"]
      }
    } else {
      return value
    }
  }

  const t: TranslateFn = (locales, args) => {
    let value = getValue(locales, args?.count)

    if (args?.params) {
      value = value.toString().replace(STR_REGEXP, function (_, variable, arg) {
        const substitute = args.params[variable]
        if (typeof substitute === "function") {
          return substitute(arg)
        } else {
          return substitute.toString()
        }
      })

      if (TAG_REGEXP.test(value)) {
        return value.split(TAG_REGEXP).map((part, i, array) => {
          // must match 2nd and 3rd param on a group of four
          // see tests for example
          if (i % 3 === 1) {
            const element = args.params[part]
            if (!isValidElement(element)) {
              throw new Error(
                "I18n received a non JSX element. Only use <tag> substitutes when variables are meant to be replaced by JSX elements.",
              )
            }
            // empty the next one, which is a child element
            const child = array[i + 1]
            array[i + 1] = ""
            return cloneElement(element, {
              ...element.props,
              key: i,
              children: child,
            })
          } else {
            return part
          }
        })
      }
    }

    return value
  }

  return {
    t,
    locale,
    locales,
    defaultLocale,
  }
}

export default useI18n
