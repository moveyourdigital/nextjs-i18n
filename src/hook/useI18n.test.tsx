import mockConsole from "jest-mock-console"
import useI18n from "./useI18n"

const i18n = {
  locale: "pt",
  locales: ["en", "pt"],
  defaultLocale: "en",
}

jest.mock("next/router", () => ({
  useRouter: () => i18n,
}))

describe("useI18n", () => {
  beforeEach(() => {
    mockConsole()
  })

  describe("t", () => {
    const { t } = useI18n()

    it.each`
      count
      ${"0"}
      ${"1"}
      ${"2"}
      ${"42"}
    `("returns the string for $count messages", ({ count }) => {
      const result = t(
        {
          pt: {
            0: 'Olá {{name}}. Não tem <link>mensagens</link> do dia <b>{{today "d/m/Y"}}</b>',
            1: 'Olá {{name}}. Tem {{count}} <link>mensagem</link> do dia <b>{{today "d/m/Y"}}</b>',
            n: 'Olá {{name}}. Tem {{count}} <link>mensagens</link> do dia <b>{{today "d/m/Y"}}</b>',
          },
          en: {
            0: 'Hi {{name}}. No <link>messages</link> for <b>{{today "m/d/Y"}}</b>',
            1: 'Hi {{name}}. You have {{count}} <link>message</link> for <b>{{today "m/d/Y"}}</b>',
            n: 'Hi {{name}}. You have {{count}} <link>messages</link> for <b>{{today "m/d/Y"}}</b>',
          },
        },
        {
          count,
          params: {
            name: "Jane Doe",
            count,
            link: <a href="/messages" />,
            b: <strong />,
            today: (format: string) => {
              const date = new Date("2021-02-28")
              return format.replace(/d|m|Y/g, (match) => {
                switch (match) {
                  case "d":
                    return date.getDate().toString()
                  case "m":
                    return (date.getMonth() + 1).toString()
                  case "Y":
                    return date.getFullYear().toString()
                }
              })
            },
          },
        },
      )

      expect(result).toMatchSnapshot()
    })
  })
})
