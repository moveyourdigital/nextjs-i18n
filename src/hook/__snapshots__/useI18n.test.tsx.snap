// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useI18n t returns the string for 0 messages 1`] = `
Array [
  "Olá Jane Doe. Não tem ",
  <a
    href="/messages"
  >
    mensagens
  </a>,
  "",
  " do dia ",
  <strong>
    28/2/2021
  </strong>,
  "",
  "",
]
`;

exports[`useI18n t returns the string for 1 messages 1`] = `
Array [
  "Olá Jane Doe. Tem 1 ",
  <a
    href="/messages"
  >
    mensagem
  </a>,
  "",
  " do dia ",
  <strong>
    28/2/2021
  </strong>,
  "",
  "",
]
`;

exports[`useI18n t returns the string for 2 messages 1`] = `"Olá Jane Doe. Tem 2 <link>mensagens</link> do dia <b>28/2/2021</b>"`;

exports[`useI18n t returns the string for 42 messages 1`] = `
Array [
  "Olá Jane Doe. Tem 42 ",
  <a
    href="/messages"
  >
    mensagens
  </a>,
  "",
  " do dia ",
  <strong>
    28/2/2021
  </strong>,
  "",
  "",
]
`;
